sympy/solvers/solvers.py

```diff
@@ -274,10 +275,6 @@ def checksol(f, symbol, sol=None, **flags):
     elif not f.is_Relational and not f:
         return True
 
-    if sol and not f.free_symbols & set(sol.keys()):
```

Comment:
why was this removed?

Reply:
It was causing inconsistency issues. To refer to my previous comment above, 
> @smichr @oscarbenjamin Do you think cases such as checksol(y, x, 2) returning None (the current behavior on master)  makes sense? I think this causes an inconsistency with something like checksol(x, x, y) (which returns False on master), > because both result in the same expression after the solution is substituted in but still have different return values.
If this is indeed an inconsistency, the following lines can be entirely removed to fix the inconsistent behavior. This  change along with sympifying the input will also fix the original issue that this PR addresses.
https://github.com/sympy/sympy/blob/9fcdf264f3310e82a7f1d159d69812cf2422dbd0/sympy/solvers/solvers.py#L278-L280

Reply:
Perhaps a better thing to do is to test the input against expectation: `not f.free_symbols - set(sol.keys())`. It is supposed to be answering False if we know that a solution does not reduce the expression to 0, None if it can't be determined, and True if it does reduce to zero. The case for args `(y, x, 2)` is an error while the case of `(x, x, y)` is False.

Reply:
I'm not sure I understand you correctly. Are you suggesting that `not f.free_symbols & set(sol.keys())` in the if condition be replaced with `f.free_symbols - set(sol.keys())`? This would not work correctly if the args are something like `(x*y, x, 0)`.
> The case for args (y, x, 2) is an error.

Do you mean that the `None` result for `(y, x, 2)` is incorrect and should be `False` instead?

Reply:
I am getting the feeling that we are turning something with a specific use case in mind (though not clearly stated) into something more general that we shouldn't (won't) need. Let's see if the following clarified possibilities make sense:

1. if a solution sets an equation to 0, return True
```
(x*y, x, 0) -> True
```
2. if a solution sets an equation to an expression that is not zero, return False
```
(cos(x)**2+sin(x)**2, x, 1) -> False
(y, x, 0) -> False (** but see below)
(x*y, x, 2) -> False
```
4.  if a solution sets an equation to a  number that can't be distinguished from zero, return None
```
(cos(x)**2+sin(x)**2 - 1, x, 1) -> None
```
(**) For the second example of 2, we might consider this a None case since there is no symbol to replace for the test to be done. But it could be the case that such an expression is just an unsimplified symbolic 0 (i.e. replace `y` with `f(y)` such that `f(y) != 0 and f(y).simplify() == 0`) so we shouldn't return False/None until the required check has been done.

Reply:
I agree with you about points 1 and 2. but I'm not sure about the example for point 4. Currently, `checksol(cos(x)**2+sin(x)**2-1, x, 1)` returns `True` because it checks the solution numerically and finds it to be sufficiently close to zero. Even if `numerical` is set to `False`, `simplify` is able to simplify `sin(1)**2+cos(1)**2-1` to 0, so `checksol` still returns `True`, which is mathematically correct. Am I understanding you correctly on point 4?
Also, entirely removing the `if sol and not f.free_symbols & set(sol.keys()):` check makes `checksol` work as you describe in points 1 and 2, so I think it's a good idea to remove it. Do you have any other suggestions in mind here?